src/boot/efi/bcd.c

```diff
@@ -0,0 +1,286 @@
+        UINT32 subkeys_offset;
+        UINT32 _pad2;
+        UINT32 n_key_values;
+        UINT32 key_values_offset;
+        UINT32 _pad3[7];
+        UINT16 key_name_len;
+        UINT16 _pad4;
+        CHAR8 key_name[];
+} _packed_ Key;
+assert_cc(offsetof(Key, sig) == 0);
+assert_cc(offsetof(Key, subkeys_offset) == 28);
+assert_cc(offsetof(Key, n_key_values) == 36);
+assert_cc(offsetof(Key, key_values_offset) == 40);
+assert_cc(offsetof(Key, key_name_len) == 72);
+assert_cc(offsetof(Key, key_name) == 76);
+
+typedef struct {
+        UINT16 sig;
+        UINT16 n_entries;
+        struct SubkeyFastEntry {
+                UINT32 key_offset;
+                CHAR8 name_hint[4];
+        } _packed_ entries[];
+} _packed_ SubkeyFast;
+assert_cc(offsetof(SubkeyFast, sig) == 0);
+assert_cc(offsetof(SubkeyFast, n_entries) == 2);
+assert_cc(offsetof(SubkeyFast, entries) == 4);
+
+typedef struct {
+        UINT16 sig;
+        UINT16 name_len;
+        UINT32 data_size;
+        UINT32 data_offset;
+        UINT32 data_type;
+        UINT32 _pad;
+        CHAR8 name[];
+} _packed_ KeyValue;
+assert_cc(offsetof(KeyValue, sig) == 0);
+assert_cc(offsetof(KeyValue, name_len) == 2);
+assert_cc(offsetof(KeyValue, data_size) == 4);
+assert_cc(offsetof(KeyValue, data_offset) == 8);
+assert_cc(offsetof(KeyValue, data_type) == 12);
+assert_cc(offsetof(KeyValue, name) == 20);
+
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name);
+
+static Key *get_subkey(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name) {
+        assert(bcd);
+        assert(name);
+
+        if ((UINT64) offset + sizeof(SubkeyFast) > bcd_len)
+                return NULL;
+
+        SubkeyFast *subkey = (SubkeyFast *) (bcd + offset);
+        if (subkey->sig != SIG_SUBKEY_FAST)
+                return NULL;
+
+        if ((UINT64) offset + offsetof(SubkeyFast, entries) + sizeof(struct SubkeyFastEntry[subkey->n_entries]) > bcd_len)
+                return NULL;
+
+        for (UINT16 i = 0; i < subkey->n_entries; i++) {
+                if (strncasecmpa(name, subkey->entries[i].name_hint, sizeof(subkey->entries[i].name_hint)) != 0)
+                        continue;
+
+                Key *key = get_key(bcd, bcd_len, subkey->entries[i].key_offset, name);
+                if (key)
+                        return key;
+        }
+
+        return NULL;
+}
+
+/* We use NUL as registry path separators for convenience. To start from the root, begin
+ * name with a NUL. Name must end with two NUL. */
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name) {
+        assert(bcd);
+        assert(name);
+
+        if ((UINT64) offset + sizeof(Key) > bcd_len)
+                return NULL;
+
+        Key *key = (Key *) (bcd + offset);
+        if (key->sig != SIG_KEY)
+                return NULL;
+
+        if ((UINT64) offset + offsetof(Key, key_name) + sizeof(CHAR8[key->key_name_len]) > bcd_len)
```

Comment:
> `sizeof(CHAR8[key->key_name_len])`

~~That's just `key->key_name_len`. Why not use that directly?~~

OK, I see this is done for consistency with `sizeof(INT32[...])` below. Makes sense.